README
NAME
    Net::Twitter::Lite - A perl interface to the Twitter API

VERSION
    version 0.12002

STOP!
    You probably want Net::Twitter::Lite::WithAPIv1_1 which has support for
    Twitter API v1.1. If you're using a service with an API compatible with
    Twitter's deprecated API v1, then you're in the right place.

SYNOPSIS
      use Net::Twitter::Lite;

      my $nt = Net::Twitter::Lite->new(
          username => $user,
          password => $password
          legacy_lists_api => 0,
      );

      my $result = eval { $nt->update('Hello, world!') };

      eval {
          my $statuses = $nt->friends_timeline({ since_id => $high_water, count => 100 });
          for my $status ( @$statuses ) {
              print "$status->{created_at} <$status->{user}{screen_name}> $status->{text}\n";
          }
      };
      warn "$@\n" if $@;

DESCRIPTION
    This module provides a perl interface to the Twitter API v1.

    It uses the same API definitions as Net::Twitter, but without the extra
    bells and whistles and without the additional dependencies. Same great
    taste, less filling.

    This module is related to, but is not part of the "Net::Twitter"
    distribution. It's API methods and API method documentation are
    generated from "Net::Twitter"'s internals. It exists for those who
    cannot, or prefer not to install Moose and its dependencies.

    You should consider upgrading to "Net::Twitter" for additional
    functionality, finer grained control over features, backwards
    compatibility with older versions of "Net::Twitter", and additional
    error handling options.

CLIENT CODE CHANGES REQUIRED
  Legacy Lists API
    Twitter re-implemented the Lists API using new endpoints and semantics.
    For backwards compatibility, this version of "Net::Twitter::Lite"
    defaults to the deprecated, legacy endpoints and semantics. It issues a
    warning if the "legacy_lists_api" option to new is not provided.

    To enable the new Lists endpoints and semantics, pass "(legacy_lists_api
    =" 0)> to "new". To disable the warning, and keep the backwards
    compatible endpoints and semantics, pass "(legacy_lists_api =" 1)> to
    "new".

    The "legacy_lists_api" option to "new" sets the default for all lists
    API method calls. You can override the default an each API call by
    passing a "-legacy_lists_api" option set to 1 or 0.

    Support for "legacy_lists_api" option will be removed in a future
    version and the option to "new" will be silently ignored.

  netrc option
    The default "apiurl" changed in version 0.08006. The change should be
    transparent to client code, unless you're using the "netrc" option. If
    so, you'll need to either update the ".netrc" entry and change the
    "machine" value from "twitter.com" to "api.twitter.com", or set either
    the "netrc" or "netrc_machine" options to "twitter.com".

        $nt = Net::Twitter::Lite->new(netrc_machine => 'twitter.com', netrc => 1);
        # -or-
        $nt = Net::Twitter::Lite->new(netrc => 'twitter.com');

  OAuth requires callback parameter
    Beginning with version 0.03, it is necessary for web applications using
    OAuth authentication to pass the "callback" parameter to
    "get_authorization_url". In the absence of a callback parameter, when
    the user authorizes the application a PIN number is displayed rather
    than redirecting the user back to your site.

MIGRATING FROM NET::TWITTER 2.x
    If you are migrating from Net::Twitter 2.12 (or an earlier version), you
    may need to make some minor changes to your application code in order to
    user Net::Twitter::Lite successfully.

    The primary difference is in error handling. Net::Twitter::Lite throws
    exceptions on error. It does not support the "get_error", "http_code",
    and "http_message" methods used in Net::Twitter 2.12 and prior versions.

    Instead of

      # DON'T!
      my $friends = $nt->friends();
      if ( $friends ) {
          # process $friends
      }

    wrap the API call in an eval block:

      # DO!
      my $friends = eval { $nt->friends() };
      if ( $friends ) {
          # process $friends
      }

    Here's a much more complex example taken from application code using
    Net::Twitter 2.12:

      # DON'T!
      my $friends = $nt->friends();
      if ( $friends ) {
          # process $friends
      }
      else {
          my $error = $nt->get_error;
          if ( ref $error ) {
              if ( ref($error) eq 'HASH' && exists $error->{error} ) {
                  $error = $error->{error};
              }
              else {
                  $error = 'Unexpected error type ' . ref($error);
              }
          }
          else {
              $error = $nt->http_code() . ": " . $nt->http_message;
          }
          warn "$error\n";
      }

    The Net::Twitter::Lite equivalent is:

      # DO!
      eval {
          my $friends = $nt->friends();
          # process $friends
      };
      warn "$@\n" if $@;
      return;

    In Net::Twitter::Lite, an error can always be treated as a string. See
    Net::Twitter::Lite::Error. The HTTP Status Code and HTTP Message are
    both available. Rather than accessing them via the Net::Twitter::Lite
    instance, you access them via the Net::Twitter::Lite::Error instance
    thrown as an error.

    For example:

      # DO!
      eval {
         my $friends = $nt->friends();
         # process $friends
      };
      if ( my $error = $@ ) {
          if ( blessed $error && $error->isa("Net::Twitter::Lite::Error)
               && $error->code() == 502 ) {
              $error = "Fail Whale!";
          }
          warn "$error\n";
      }

  Unsupported Net::Twitter 2.12 options to "new"
    Net::Twitter::Lite does not support the following Net::Twitter 2.12
    options to "new". It silently ignores them:

    no_fallback
        If Net::Twitter::Lite is unable to create an instance of the class
        specified in the "useragent_class" option to "new", it dies, rather
        than falling back to an LWP::UserAgent object. You really don't want
        a failure to create the "useragent_class" you specified to go
        unnoticed.

    twittervision
        Net::Twitter::Lite does not support the TwitterVision API. Use
        Net::Twitter, instead, if you need it.

    skip_arg_validation
        Net::Twitter::Lite does not API parameter validation. This is a
        feature. If Twitter adds a new option to an API method, you can use
        it immediately by passing it in the HASH ref to the API call.

        Net::Twitter::Lite relies on Twitter to validate its own parameters.
        An appropriate exception will be thrown if Twitter reports a
        parameter error.

    die_on_validation
        See "skip_arg_validation". If Twitter returns an bad parameter
        error, an appropriate exception will be thrown.

    arrayref_on_error
        This option allowed the following idiom in Net::Twitter 2.12:

          # DON'T!
          for my $friend ( @{ $nt->friends() } ) {
             # process $friend
          }

        The equivalent Net::Twitter::Lite code is:

          # DO!
          eval {
              for my $friend ( @{ $nt->friends() } ) {
                  # process $friend
              }
          };

  Unsupported Net::Twitter 2.12 methods
    clone
        The "clone" method was added to Net::Twitter 2.x to allow safe error
        handling in an environment where concurrent requests are handled,
        for example, when using LWP::UserAgent::POE as the
        "useragent_class". Since Net::Twitter::Lite throws exceptions
        instead of stashing them in the Net::Twitter::Lite instance, it is
        safe in a current request environment, obviating the need for
        "clone".

    get_error
    http_code
    http_message
        These methods are replaced by Net::Twitter::Lite::Error. An instance
        of that class is thrown errors are encountered.

METHODS AND ARGUMENTS
    new This constructs a "Net::Twitter::Lite" object. It takes several
        named parameters, all of them optional:

        username
            This is the screen name or email used to authenticate with
            Twitter. Use this option for Basic Authentication, only.

        password
            This is the password used to authenticate with Twitter. Use this
            option for Basic Authentication, only.

        consumer_key
            A string containing the OAuth consumer key provided by Twitter
            when an application is registered. Use this option for OAuth
            authentication, only.

        consumer_secret
            A string containing the OAuth consumer secret. Use this option
            for OAuth authentication, only. the "OAuth" trait is included.

        oauth_urls
            A HASH ref of URLs to be used with OAuth authentication.
            Defaults to:

              {
                  request_token_url => "http://twitter.com/oauth/request_token",
                  authorization_url => "http://twitter.com/oauth/authorize",
                  access_token_url  => "http://twitter.com/oauth/access_token",
                  xauth_url         => "https://twitter.com/oauth/access_token",
              }

        clientname
            The value for the "X-Twitter-Client-Name" HTTP header. It
            defaults to "Perl Net::Twitter::Lite".

        clientver
            The value for the "X-Twitter-Client-Version" HTTP header. It
            defaults to current version of the "Net::Twitter::Lite" module.

        clienturl
            The value for the "X-Twitter-Client-URL" HTTP header. It
            defaults to the search.cpan.org page for the
            "Net::Twitter::Lite" distribution.

        useragent_class
            The "LWP::UserAgent" compatible class used internally by
            "Net::Twitter::Lite". It defaults to "LWP::UserAgent". For POE
            based applications, consider using "LWP::UserAgent::POE".

        useragent_args
            An HASH ref of arguments to pass to constructor of the class
            specified with "useragent_class", above. It defaults to {} (an
            empty HASH ref).

        useragent
            The value for "User-Agent" HTTP header. It defaults to
            "Net::Twitter::Lite/0.11002 (Perl)".

        source
            The value used in the "source" parameter of API method calls. It
            is currently only used in the "update" method in the REST API.
            It defaults to "twitterpm". This results in the text "from
            Net::Twitter" rather than "from web" for status messages posted
            from "Net::Twitter::Lite" when displayed via the Twitter web
            interface. The value for this parameter is provided by Twitter
            when a Twitter application is registered. See
            <http://apiwiki.twitter.com/FAQ#HowdoIget%E2%80%9CfromMyApp%E2%8
            0%9DappendedtoupdatessentfrommyAPIapplication>.

        apiurl

Changes
Revision history for Net::Twitter::Lite

{{$NEXT}}

0.12008  2016-12-05 17:21:25 PST
    - modifed example/oauth_webapp.pl (issue 18)
    - added upload_media; deprecated update_with_media

0.12007 2016-11-18
    - depend on LWP::Protocol::https instead of Crypt::SSLeay
    - make SSL the default for Net::Twitter::Lite::WithAPIv1_1

0.12006 2014-01-24
    - Added WrapResult: return both the Twitter result and HTTP::Response

0.12005 2014-01-17
    - Twitter now requires SSL connections

0.12004 2013-05-03
    - restore list_subscriptions

0.12003 2013-04-25
    - Added list_ownerships

0.12002 2013-03-11
    - 30 second delay in OAuth requests fixed with Net::HTTP 6.06

0.12001 2013-03-06
    - Fixed tests (Net::OAuth is optional) (closses issue #7)
    - Updated examples to use Twitter API v1.1

0.12000 2013-03-04
    - URI encode POSTDATA to match Twitter's more stringent (non-standard) requirements
    - created Net::Twitter::Lite::WithAPIv1_1 for Twitter API v1.1 support

0.11002 2012-06-27
    - rebuilt from Net::Twitter 3.18003 to fix replies, mentions path

0.11001 2012-05-08
    - CPAN release

0.11000_01 2012-04-24
    - functional parity with Net::Twitter 3.18002
    - Replaced JSON::Any with JSON
    - new lists methods, legacy_lists_api option for back compat
    - update_with_media and other new API methods ported from Net::Twitter

0.10004 2011-03-28
	- ported from Net::Twitter 3.16000
	- fixed: methods requiring HTTP DELETE failed with OAuth
    - doc fix: use verifier in request_access_token call
    - always use SSL for token exchange

0.10003 2010-05-27
    - fixed ssl test: skip_all if Crypt::SSLeay is not available

0.10002 2010-05-27
    - fixed: set correct url to https when ssl is specified RT#57889

0.10001 2010-05-26
    - fixed "plan twice" error in unicode test

0.10000 2010-05-21
    - production release, ported from Net::Twitter 3.13003

0.09002_03 2010-05-13
    - ported from Net::Twitter 3.13002_03
    - handle DELETE HTTP method (for Lists API)

0.09002_02 2010-05-12
    - ported from Net::Twitter 3.13002_02
    - Added Lists API support

0.09001 2010-05-10
    - removed Try::Tiny from unicode test to avoid an additional dependency
    - Fixed unicode test, aburch (Ansgar Burchardt)

0.09000 2010-05-09
    - New API methods (from Net::Twitter): friendships_incoming,
      friendships_outgoing, geo_id, retweeted_by, retweeted_by_ids,
      reverse_geocode.
    - Added XAuth support (method xauth)
    - Fixed: Net::Twitter::Lite::Error - needed overload fallback => 1
    - Fixed: utf8 encoding error for latin1 using Basic Authentication (ported from Net::Twitter)

0.08006 2009-12-15
    - Ported from Net::Twitter:
      - netrc/netrc_machine changes from Net::Twitter
      - trends_available/trends_location API methods
      - documented lat/long parameters to update
      - added users_search/find_people API method
      - updated apiurl default to http://api.twitter.com/1
    - Fixed: retweeted_{to,of}_me API URLs (RT#52784)

0.08005 2009-11-02
    - Return user_id and screen_name from request_access token (ported from Net::Twitter)

0.08004 2009-10-22
    - Fixed: cached request_token resulted in stale authentication/autorization urls
    - Added better runtime error message when SSL support is missing

0.08003 2009-10-13
    - Added new API method: report_spam

0.08002 2009-10-12
    - ported support for multipart/from-data posts, used by update_profile_image and
      update_profile_background_image, from Net::Twitter
    - Improved error message when Net::OAuth is not install or version is inadequate
    - Changed nonce generation to remove dependency on Digest::SHA
    - Doc patch: created_at, not time (thanks to Blair Christensen RT#50313)

0.08001 2009-09-10
    - URI 1.35 causes utf8 encoding problems; require 1.40 (thanks to Dan Boger, @zigdon)
    - Removed deprecated is_authorized from examples in pod (thanks to Nigel Metheringham)

0.08000 2009-08-27
    - Ported get_authentication_url from Net::Twitter
    - Work around perl bug requiring encoded hash keys when client uses "use utf8"

0.07000 2009-08-14
    - Removed JSON::DWIW as a handler (no support for JSON::Any's utf8 option)

0.06002 2009-07-29
    - Fix: spurious OAuth signature failures (ported from Net::Twitter)

0.06001 2009-07-28
    - Fix: OAuth / unicode conflict

0.06000 2009-07-28
    - Bug fix: OAuth signatures on POST requests (Galen Huntington)
    - Unicode fixes

0.05000 2009-07-09
    - Added ssl and netrc arguments to new

0.04001 2009-07-07
    - Added documentation for the required callback param to
      get_authorization_url
    - New nonce algorithm to eliminate potential duplicates on forked
      processes

0.04000 2009-06-27
    - Added support for the "authenticate" parameter to API methods

    - Twitter API update:
        - Added screen_name and user_id parameters to new_direct_message
        - Added show_friendship method (friendships/show.json)

0.03001 2009-06-20
    - Fixed Basic Auth regression error

0.03000 2009-06-20
    - Added OAuth 1.0a support
    - Added examples using OAuth

0.02002 2009-06-11
    - Reverted the 'forceauth' feature; it requires too recent LWP::UserAgent

0.02000 2009-06-07
    - Added support for the saved_searches API methods

0.01001 2009-06-06
    - Fixed: accept extra args as a hashref (search behaved this way in 2.12)

0.01000 2009-06-06
    - non-Moose variation of Net::Twitter


lib/Net/Twitter/Lite.pm
package Net::Twitter::Lite;
our $VERSION = '0.12008';
use 5.005;
use warnings;
use strict;

=head1 NAME

Net::Twitter::Lite - A perl library for Twitter's API v1

=cut

use Carp;
use URI::Escape;
use JSON;
use HTTP::Request::Common;
use Net::Twitter::Lite::Error;
use Encode qw/encode_utf8/;
use Net::Twitter::Lite::WrapResult;

sub twitter_api_def_from           () { 'Net::Twitter::Lite::API::V1' }
sub _default_api_url               () { 'http://api.twitter.com/1'    }
sub _default_searchapiurl          () { 'http://search.twitter.com'   }
sub _default_search_trends_api_url () { 'http://api.twitter.com/1'    }
sub _default_lists_api_url         () { 'http://api.twitter.com/1'    }

my $json_handler = JSON->new->utf8;

sub new {
    my ($class, %args) = @_;

    $class->can('verify_credentials') || $class->build_api_methods;

    my $netrc = delete $args{netrc};
    my $new = bless {
        apiurl                => $class->_default_api_url,
        searchapiurl          => $class->_default_searchapiurl,
        search_trends_api_url => $class->_default_search_trends_api_url,
        lists_api_url         => $class->_default_lists_api_url,
        apirealm   => 'Twitter API',
        $args{identica} ? ( apiurl => 'http://identi.ca/api' ) : (),
        useragent  => (ref $class || $class) . "/$VERSION (Perl)",
        clientname => (ref $class || $class),
        clientver  => $VERSION,
        clienturl  => 'http://search.cpan.org/dist/Net-Twitter-Lite/',
        source     => 'twitterpm',
        useragent_class => 'LWP::UserAgent',
        useragent_args  => {},
        oauth_urls => {
            request_token_url  => "https://api.twitter.com/oauth/request_token",
            authentication_url => "https://api.twitter.com/oauth/authenticate",
            authorization_url  => "https://api.twitter.com/oauth/authorize",
            access_token_url   => "https://api.twitter.com/oauth/access_token",
            xauth_url          => "https://api.twitter.com/oauth/access_token",
        },
        netrc_machine => 'api.twitter.com',
        %args
    }, $class;

    unless ( exists $new->{legacy_lists_api} ) {
        $new->{legacy_lists_api} = 1;
        carp
"For backwards compatibility @{[ __PACKAGE__ ]} uses the deprecated Lists API
endpoints and semantics. This default will be changed in a future version.
Please update your code to use the new lists semantics and pass
(legacy_lists_api => 0) to new.

You can disable this warning, and keep backwards compatibility by passing
(legacy_lists_api => 1) to new. Be warned, however, that support for the
legacy endpoints will be removed in a future version and the default will
change to (legacy_lists_api => 0).";

    }

    if ( delete $args{ssl} ) {
        $new->{$_} =~ s/^http:/https:/
            for qw/apiurl searchapiurl search_trends_api_url lists_api_url/;
    }

    # get username and password from .netrc
    if ( $netrc ) {
        eval { require Net::Netrc; 1 }
            || croak "Net::Netrc is required for the netrc option";

        my $host = $netrc eq '1' ? $new->{netrc_machine} : $netrc;
        my $nrc = Net::Netrc->lookup($host)
            || croak "No .netrc entry for $host";

        @{$new}{qw/username password/} = $nrc->lpa;
    }

    $new->{ua} ||= do {
        eval "use $new->{useragent_class}";
        croak $@ if $@;

        $new->{useragent_class}->new(%{$new->{useragent_args}});
    };

    $new->{ua}->agent($new->{useragent});
    $new->{ua}->default_header('X-Twitter-Client'         => $new->{clientname});
    $new->{ua}->default_header('X-Twitter-Client-Version' => $new->{clientver});
    $new->{ua}->default_header('X-Twitter-Client-URL'     => $new->{clienturl});
    $new->{ua}->env_proxy;

    $new->{_authenticator} = exists $new->{consumer_key}
                           ? '_oauth_authenticated_request'
                           : '_basic_authenticated_request';

    $new->credentials(@{$new}{qw/username password/})
        if exists $new->{username} && exists $new->{password};

    return $new;
}

sub credentials {
    my $self = shift;
    my ($username, $password) = @_;

    croak "exected a username and password" unless @_ == 2;
    croak "OAuth authentication is in use"  if exists $self->{consumer_key};

    $self->{username} = $username;
    $self->{password} = $password;

    my $uri = URI->new($self->{apiurl});
    my $netloc = join ':', $uri->host, $uri->port;

    $self->{ua}->credentials($netloc, $self->{apirealm}, $username, $password);
}

# This is a hack. Rather than making Net::OAuth an install requirement for
# Net::Twitter::Lite, require it at runtime if any OAuth methods are used.  It
# simply returns the string 'Net::OAuth' after successfully requiring
# Net::OAuth.
sub _oauth {
    my $self = shift;

    return $self->{_oauth} ||= do {
        eval "use Net::OAuth 0.25";
        croak "Install Net::OAuth 0.25 or later for OAuth support" if $@;

        eval '$Net::OAuth::PROTOCOL_VERSION = Net::OAuth::PROTOCOL_VERSION_1_0A';
        die $@ if $@;

        'Net::OAuth';
    };
}

# simple check to see if we have access tokens; does not check to see if they are valid
sub authorized {
    my $self = shift;

    return defined $self->{access_token} && $self->{access_token_secret};
}

# get the athorization or authentication url
sub _get_auth_url {
    my ($self, $which_url, %params ) = @_;

    $self->_request_request_token(%params);

    my $uri = $self->$which_url;
    $uri->query_form(oauth_token => $self->request_token);
    return $uri;
}

# get the authentication URL from Twitter
sub get_authentication_url { return shift->_get_auth_url(authentication_url => @_) }

# get the authorization URL from Twitter
sub get_authorization_url { return shift->_get_auth_url(authorization_url => @_) }

# common portion of all oauth requests
sub _make_oauth_request {
    my ($self, $type, %params) = @_;

    my $request = $self->_oauth->request($type)->new(
        version          => '1.0',
        consumer_key     => $self->{consumer_key},
        consumer_secret  => $self->{consumer_secret},
        request_method   => 'GET',
        signature_method => 'HMAC-SHA1',
        timestamp        => time,
        nonce            => time ^ $$ ^ int(rand 2**32),
        %params,
    );

    $request->sign;

    return $request;
}

# called by get_authorization_url to obtain request tokens
sub _request_request_token {
    my ($self, %params) = @_;

    my $uri = $self->request_token_url;
    $params{callback} ||= 'oob';
    my $request = $self->_make_oauth_request(
        'request token',
        request_url => $uri,
        %params,
    );

    my $res = $self->{ua}->get($request->to_url);
    die "GET $uri failed: ".$res->status_line
        unless $res->is_success;

    # reuse $uri to extract parameters from the response content
    $uri->query($res->content);
    my %res_param = $uri->query_form;

    $self->request_token($res_param{oauth_token});
    $self->request_token_secret($res_param{oauth_token_secret});
}

# exchange request tokens for access tokens; call with (verifier => $verifier)
sub request_access_token {
    my ($self, %params ) = @_;

    my $uri = $self->access_token_url;
    my $request = $self->_make_oauth_request(
        'access token',
        request_url => $uri,
        token       => $self->request_token,
        token_secret => $self->request_token_secret,
        %params, # verifier => $verifier
    );

    my $res = $self->{ua}->get($request->to_url);
    die "GET $uri failed: ".$res->status_line
        unless $res->is_success;

    # discard request tokens, they're no longer valid
    delete $self->{request_token};
    delete $self->{request_token_secret};

    # reuse $uri to extract parameters from content
    $uri->query($res->content);
    my %res_param = $uri->query_form;

    return (
        $self->access_token($res_param{oauth_token}),
        $self->access_token_secret($res_param{oauth_token_secret}),
        $res_param{user_id},
        $res_param{screen_name},
    );
}

# exchange username and password for access tokens
sub xauth {
    my ( $self, $username, $password ) = @_;

    my $uri = $self->xauth_url;
    my $request = $self->_make_oauth_request(
        'XauthAccessToken',
        request_url     => $uri,
        x_auth_username => $username,
        x_auth_password => $password,
        x_auth_mode     => 'client_auth',
    );

    my $res = $self->{ua}->get($request->to_url);
    die "GET $uri failed: ".$res->status_line
        unless $res->is_success;

    # reuse $uri to extract parameters from content
    $uri->query($res->content);
    my %res_param = $uri->query_form;

    return (
        $self->access_token($res_param{oauth_token}),
        $self->access_token_secret($res_param{oauth_token_secret}),
        $res_param{user_id},
        $res_param{screen_name},
    );
}

# common call for both Basic Auth and OAuth
sub _authenticated_request {
    my $self = shift;

    my $authenticator = $self->{_authenticator};
    $self->$authenticator(@_);
}

sub _encode_args {
    my $args = shift;

    # Values need to be utf-8 encoded.  Because of a perl bug, exposed when
    # client code does "use utf8", keys must also be encoded.
    # see: http://www.perlmonks.org/?node_id=668987
    # and: http://perl5.git.perl.org/perl.git/commit/eaf7a4d2
    return { map { utf8::upgrade($_) unless ref($_); $_ } %$args };
}

sub _oauth_authenticated_request {
    my ($self, $http_method, $uri, $args, $authenticate) = @_;

    delete $args->{source}; # not necessary with OAuth requests

    my $content_type = delete $args->{-content_type} || '';
    my $is_multipart = $content_type eq 'form-data' || grep { ref } %$args;

    my $msg;
    if ( $authenticate && $self->authorized ) {
        local $Net::OAuth::SKIP_UTF8_DOUBLE_ENCODE_CHECK = 1;

        my $request = $self->_make_oauth_request(
            'protected resource',
            request_url    => $uri,
            request_method => $http_method,
            token          => $self->access_token,
            token_secret   => $self->access_token_secret,
            extra_params   => $is_multipart ? {} : $args,
        );

        if ( $http_method =~ /^(?:GET|DELETE)$/ ) {
            $msg = HTTP::Request->new($http_method, $request->to_url);
        }
        elsif ( $http_method eq 'POST' ) {
            $msg = $is_multipart
                 ? POST($request->request_url,
                        Authorization => $request->to_authorization_header,
                        Content_Type  => 'form-data',
                        Content       => [ %$args ],
                   )
                 : POST($$uri, Content => $request->to_post_body)
                 ;
        }
        else {
            croak "unexpected http_method: $http_method";
        }
    }
    elsif ( $http_method eq 'GET' ) {
        $uri->query_form($args);
        $args = {};
        $msg = GET($uri);
    }
    elsif ( $http_method eq 'POST' ) {
        my $encoded_args = { %$args };
        _encode_args($encoded_args);
        $msg = $self->_mk_post_msg($uri, $args);
    }
    else {
        croak "unexpected http_method: $http_method";
    }

    return $self->{ua}->request($msg);
}

sub _basic_authenticated_request {
    my ($self, $http_method, $uri, $args, $authenticate) = @_;

    _encode_args($args);

    my $msg;
    if ( $http_method =~ /^(?:GET|DELETE)$/ ) {
        $uri->query_form($args);
        $msg = HTTP::Request->new($http_method, $uri);
    }
    elsif ( $http_method eq 'POST' ) {
        $msg = $self->_mk_post_msg($uri, $args);
    }
    else {
        croak "unexpected HTTP method: $http_method";
    }

    if ( $authenticate && $self->{username} && $self->{password} ) {
        $msg->headers->authorization_basic(@{$self}{qw/username password/});
    }

    return $self->{ua}->request($msg);
}

sub _mk_post_msg {
    my ($self, $uri, $args) = @_;

    if ( grep { ref } values %$args ) {
        # if any of the arguments are (array) refs, use form-data
        return POST($uri, Content_Type => 'form-data', Content => [ %$args ]);
    }
    else {
        # There seems to be a bug introduced by Twitter about 2013-02-25: If
        # post arguments are uri encoded exactly the same way the OAuth spec
        # requires base signature string encoding, Twitter chokes and throws a
        # 401.  This seems to be a violation of the OAuth spec on Twitter's
        # part. The specifically states the more stringent URI encoding is for
        # consistent signature generation and *only* applies to encoding the
        # base signature string and Authorization header.

        my @pairs;
        while ( my ($k, $v) = each %$args ) {
            push @pairs, join '=', map URI::Escape::uri_escape_utf8($_, '^A-Za-z0-9\-\._~'), $k, $v;
        }

        my $content = join '&', @pairs;
        return POST($uri, Content => $content);
    }
}

sub build_api_methods {
    my $class = shift;

    my $api_def_module = $class->twitter_api_def_from;
    eval "require $api_def_module" or die $@;
    my $api_def = $api_def_module->api_def;

    my $with_url_arg = sub {
        my ($path, $args) = @_;

        if ( defined(my $id = delete $args->{id}) ) {
            $path .= uri_escape($id);
        }
        else {
            chop($path);
        }
        return $path;
    };

    while ( @$api_def ) {
        my $api = shift @$api_def;
        my $api_name = shift @$api;
        my $methods = shift @$api;

        for my $method ( @$methods ) {
            my $name    = shift @$method;
            my %options = %{ shift @$method };

            my ($arg_names, $path) = @options{qw/required path/};
            $arg_names = $options{params}
                if @$arg_names == 0 && @{$options{params}} == 1;

            my $modify_path = $path =~ s,/id$,/, ? $with_url_arg : sub { $_[0] };

            my $code = sub {
                my $self = shift;

                # copy callers args since we may add ->{source}
                my $args = ref $_[-1] eq 'HASH' ? { %{pop @_} } : {};
                if ( my $content_type = $options{content_type} ) {
                    $args->{-content_type} = $options{content_type};
                }

                if ( (my $legacy_method = $self->can("legacy_$name")) && (
                        exists $$args{-legacy_lists_api}
                            ? delete $$args{-legacy_lists_api}
                            : $self->{legacy_lists_api} ) ) {
                    return $self->$legacy_method(@_, $args);
                }

                # just in case it's included where it shouldn't be:
                delete $args->{-legacy_lists_api};

                croak sprintf "$name expected %d args", scalar @$arg_names
                    if @_ > @$arg_names;

                # promote positional args to named args
                for ( my $i = 0; @_; ++$i ) {
                    my $param = $arg_names->[$i];
                    croak "duplicate param $param: both positional and named"
                        if exists $args->{$param};

                    $args->{$param} = shift;
                }

                $args->{source} ||= $self->{source} if $options{add_source};

                my $authenticate = exists $args->{authenticate}
                    ? delete $args->{authenticate}
                    : $options{authenticate};

                # promote boolean parameters
                for my $boolean_arg ( @{ $options{booleans} } ) {
                    if ( exists $args->{$boolean_arg} ) {
                        next if $args->{$boolean_arg} =~ /^true|false$/;
                        $args->{$boolean_arg} = $args->{$boolean_arg} ? 'true' : 'false';
                    }
                }

                # Workaround Twitter bug: any value passed for skip_user is treated as true.
                # The only way to get 'false' is to not pass the skip_user at all.
                delete $args->{skip_user} if exists $args->{skip_user}
                    && $args->{skip_user} eq 'false';

                # replace placeholder arguments
                my $local_path = $path;

                # remove optional trailing id
                $local_path =~ s,/:id$,, unless exists $args->{id};
                $local_path =~ s/:(\w+)/delete $args->{$1}
                    or croak "required arg '$1' missing"/eg;

                # stringify lists
                for ( qw/screen_name user_id/ ) {
                    $args->{$_} = join(',' => @{ $args->{$_} })
                        if ref $args->{$_} eq 'ARRAY';
                }

                my $uri = URI->new($self->{$options{base_url_method}}
                    . "/$local_path.json");

                return $self->_parse_result(
                    $self->_authenticated_request(
                        $options{method}, $uri, $args, $authenticate
                    )
                );
            };

            no strict 'refs';
            $name = $_, *{"$class\::$_"} = $code for $name, @{$options{aliases}};
        }
    }

    # catch expected error and promote it to an undef
    for ( qw/list_members is_list_member list_subscribers is_list_subscriber
            legacy_list_members legacy_is_list_member legacy_list_subscribers legacy_is_list_subscriber/ ) {
        my $orig = $class->can($_) or next;

        my $code = sub {
            my $r = eval { $orig->(@_) };
            if ( $@ ) {
                return if $@ =~ /The specified user is not a (?:memb|subscrib)er of this list/;

                die $@;
            }

            return $r;
        };

        no strict 'refs';
        no warnings 'redefine';
        *{"$class\::$_"} = $code;
    }

    # OAuth token accessors
    for my $method ( qw/
                access_token
                access_token_secret
                request_token
                request_token_secret
            / ) {
        no strict 'refs';
        *{"$class\::$method"} = sub {
            my $self = shift;

            $self->{$method} = shift if @_;
            return $self->{$method};
        };
    }

    # OAuth url accessors
    for my $method ( qw/
                request_token_url
                authentication_url
                authorization_url
                access_token_url
                xauth_url
            / ) {
        no strict 'refs';
        *{"$class\::$method"} = sub {
            my $self = shift;

            $self->{oauth_urls}{$method} = shift if @_;
            return URI->new($self->{oauth_urls}{$method});
        };
    }

}

sub _from_json {
    my ($self, $json) = @_;

    return eval { $json_handler->decode($json) };
}

sub _parse_result {
    my ($self, $res) = @_;

    # workaround for Laconica API returning bools as strings
    # (Fixed in Laconi.ca 0.7.4)
    my $content = $res->content;
    $content =~ s/^"(true|false)"$/$1/;

    my $obj = $self->_from_json($content);

    # Twitter sometimes returns an error with status code 200
    if ( $obj && ref $obj eq 'HASH' && exists $obj->{error} ) {
        die Net::Twitter::Lite::Error->new(twitter_error => $obj, http_response => $res);
    }

    if ( $res->is_success && defined $obj ) {
        if ( $self->{wrap_result} ) {
            $obj = Net::Twitter::Lite::WrapResult->new($obj, $res);
        }
        return $obj;
    }

    my $error = Net::Twitter::Lite::Error->new(http_response => $res);
    $error->twitter_error($obj) if ref $obj;

    die $error;
}

1;
